fix(bitcoind_rpc): properly handle reorgs in `FilterIter`

LagginTimes opened the pull request:
Replaces #1909.
Fixes #1848.

### Description

This patch enhances `FilterIter` to robustly detect and handle chain reorgs. Reorgs are detected by checking if the block hash at the current height has changed, or if the block's `prev_blockhash` no longer matches the hash of the block at the previous height. This allows detection of reorgs that occur either at the current height or one block prior.

When a reorg is detected, the iterator rewinds up to `MAX_REORG_DEPTH` blocks to locate the most recent common ancestor. It then resets its internal state and resumes scanning from the fork point on the new chain.

This update also includes tests that simulate reorgs — including back-to-back reorgs at the same height — and verifies that `FilterIter` correctly switches to the new chain and yields accurate events.

### Changelog notice

* Improve `FilterIter` to detect and handle reorgs at the current height or its parent.

### Checklists

#### All Submissions:

* [x] I've signed all my commits
* [x] I followed the [contribution guidelines](https://github.com/bitcoindevkit/bdk/blob/master/CONTRIBUTING.md)
* [x] I ran `cargo +nightly fmt` and `cargo clippy` before committing

#### New Features:

* [ ] I've added tests for the new feature
* [ ] I've added docs for the new feature

#### Bugfixes:

* [ ] This pull request breaks the existing API
* [x] I've added tests to reproduce the issue which are now passing
* [x] I'm linking the issue being fixed by this PR

Musab1258 committed:
test(bitcoind_rpc): update `FilterIter` tests

Co-authored-by: valued mammal <valuedmammal@protonmail.com>

LagginTimes committed:
fix(bitcoind_rpc): `FilterIter` detects reorgs

Co-authored-by: valued mammal <valuedmammal@protonmail.com>

LagginTimes committed:
test(bitcoind_rpc): tests for reorg between `next()` and MAX_REORG_DEPTH

ValuedMammal commented on crates/bitcoind_rpc/src/bip158.rs:
```diff
@@ -69,22 +78,14 @@ impl<'c, C: RpcApi> FilterIter<'c, C> {
+
+    /// Insert a (non-matching) block height and hash into the blocks map.
+    fn insert_block(&mut self, height: Height, hash: BlockHash) {
+        self.blocks.insert(height, hash);
```
🔧  We can get rid of the helpers `insert_block` and `get_block_hash` and simply call this code inline.

ValuedMammal commented on crates/bitcoind_rpc/src/bip158.rs:
```diff
@@ -93,33 +94,18 @@ impl<'c, C: RpcApi> FilterIter<'c, C> {
+        let header = self.client.get_block_header_info(&tip_hash)?;
         let tip_height = header.height as u32;
-        if self.height >= tip_height {
+        // Allow returning tip if we're exactly at it. Return `None`` if we've already scanned past.
+        if self.height > tip_height {
```
📌  Since this is a logical change I think the documentation for `get_tip` should be reworded slightly.

ValuedMammal commented on crates/bitcoind_rpc/src/bip158.rs:
```diff
@@ -171,27 +154,84 @@ impl<C: RpcApi> Iterator for FilterIter<'_, C> {
+            if self.height > self.stop {
+                return Ok(None);
+            }
+            // Fetch next filter.
```
```suggestion
            // Fetch next header.
```
<hr>

📌 In the future `FilterIter` should store and return the headers, that way the caller can collect a chain update into a `LocalChain<Header>`.

ValuedMammal commented on crates/bitcoind_rpc/src/bip158.rs:
```diff
@@ -171,27 +154,84 @@ impl<C: RpcApi> Iterator for FilterIter<'_, C> {
+
+            // Detect and resolve reorgs: either block at height changed, or its parent changed.
+            let stored_hash = self.blocks.get(&height).copied();
+            let prev_hash = self.blocks.get(&(height - 1)).copied();
```
🔧 This is to avoid calling `get` with the underflow height in case `height` is 0.

```suggestion
            let prev_hash = height
                .checked_sub(1)
                .and_then(|height| self.blocks.get(&height).copied());
```

ValuedMammal commented on crates/bitcoind_rpc/tests/test_filter_iter.rs:
```diff
@@ -163,3 +163,421 @@ fn filter_iter_error_no_scripts() -> anyhow::Result<()> {
+            assert_eq!(block.block_hash(), hash_104);
+            assert!(block.txdata.iter().any(|tx| tx.compute_txid() == txid_a));
+        }
+        _ => panic!("Expected relevant tx at block A 102"),
```
```suggestion
        _ => panic!("Expected relevant tx at block A 104"),
```

ValuedMammal commented on crates/bitcoind_rpc/tests/test_filter_iter.rs:
```diff
@@ -163,3 +163,421 @@ fn filter_iter_error_no_scripts() -> anyhow::Result<()> {
+    assert!(
+        final_update.is_none(),
+        "We didn't instantiate FilterIter with a checkpoint"
+    );
```
📌 It will be better to instantiate `FilterIter` with a checkpoint and check the blocks in the returned chain update as well.

ValuedMammal reviewed (commented):
Thanks for picking this up @LagginTimes

ValuedMammal reviewed (changes requested):
It's missing a test for max reorg depth.

LagginTimes commented:
> It's missing a test for max reorg depth.

The test for this is still WIP but I pushed out the changes that resolved the other open comments.

ValuedMammal commented on crates/bitcoind_rpc/src/bip158.rs:
```diff
@@ -171,27 +148,86 @@ impl<C: RpcApi> Iterator for FilterIter<'_, C> {
+
+                // Update self.height so we reprocess this height
+                self.height = height;
+            }
```
```diff
             // Reorg detected, rewind to last known-good ancestor.
             if reorg_detected {
                 self.blocks.split_off(&height);
-                self.matched = self.matched.split_off(&height);
+                self.matched.split_off(&height);
 
                 let mut reorg_depth = 0;
                 loop {
                     if reorg_depth >= Self::MAX_REORG_DEPTH || height == 0 {
                         return Err(Error::ReorgDepthExceeded);
                     }
 
                     height = height.saturating_sub(1);
                     hash = self.client.get_block_hash(height as _)?;
                     header = self.client.get_block_header(&hash)?;
 
                     let prev_height = height.saturating_sub(1);
-                    let prev_hash = self.client.get_block_hash(prev_height as _)?;
-                    self.blocks.insert(prev_height, prev_hash);
-
                     if let Some(prev_hash) = self.blocks.get(&prev_height) {
                         if header.prev_blockhash == *prev_hash {
                             break;
                         }
                     }
 
                     reorg_depth += 1;
                 }
-
-                // Update self.height so we reprocess this height
-                self.height = height;
             }

```

ValuedMammal commented on crates/bitcoind_rpc/src/bip158.rs:
```diff
@@ -171,27 +148,86 @@ impl<C: RpcApi> Iterator for FilterIter<'_, C> {
+            // Reorg detected, rewind to last known-good ancestor.
+            if reorg_detected {
+                self.blocks.split_off(&height);
+                self.matched = self.matched.split_off(&height);
```
Looks like you're using `split_off` in two different ways. If I understand the logic, I think we want to drop the return value, like you have on L178.

ValuedMammal replied:
Maybe move the `split_off` until after we've found the rewind height - so we don't hold on to any stale entries.

ValuedMammal commented on crates/bitcoind_rpc/src/bip158.rs:
```diff
@@ -171,27 +148,86 @@ impl<C: RpcApi> Iterator for FilterIter<'_, C> {
+
+                    let prev_height = height.saturating_sub(1);
+                    let prev_hash = self.client.get_block_hash(prev_height as _)?;
+                    self.blocks.insert(prev_height, prev_hash);
```
This looks problematic because we do the `header.prev_blockhash` check immediately following, which will most likely short-circuit the loop rather than rewind back to the last known good ancestor.

ValuedMammal reviewed (dismissed):
ACK c2ae4da2ca59ee05e14a99ec7fba0297bd461695

evanlinjin commented on crates/bitcoind_rpc/src/bip158.rs:
```diff
@@ -69,57 +78,28 @@ impl<'c, C: RpcApi> FilterIter<'c, C> {
     ///
-    /// Returns `None` if the remote height is not strictly greater than the height of this
-    /// [`FilterIter`].
+    /// Returns `None` if the remote height is less than the height of this [`FilterIter`].
     pub fn get_tip(&mut self) -> Result<Option<BlockId>, Error> {
```
Maybe a little out of scope for this PR, but the documentation here is a little lacking.

`get_tip`, as I understand, serves the following purposes:

1. Initialize the `FilerIter`. Without calling it, `FilterIter` returns nothing.
2. If it returns `None`, it means we are up to date with remote.

Is this the case?

If so, purpose `2.` is not satisfied because reorgs are not detected by `get_tip`.

@ValuedMammal are you able to provide some clarity, thanks!

ValuedMammal replied:
`get_tip` is used to set the start and stop height of a scan and is only necessary to call once per instance of FilterIter. During review I was going to mention that I think `get_tip` should actually clear `self.blocks`, because I didn't anticipate the user interleaving calls to `get_tip` and `next`.

evanlinjin replied:
@ValuedMammal Thank you for the update.

The internal implementation of `FilterIter` implies that it is a single-use structure (construct once, call `.next` till end). However, the API design of it implies something different.

What do you think about changing the API to properly represent a "single-use" workflow? I.e. I would change the following:

* Remove or unexpose `add_spk{s}`.
* Change constructor methods to also take in spks. Return `None` if there are no spks provided.
* Make sure `get_tip` logic is run on the first call to `next`.
* Rename `get_tip` to `init` which would return the target height.

(Separate PR of course)

Let me know what you think.

LagginTimes replied:
I like this idea, and can make a follow up PR to implement this.

notmandatory replied:
I agree it makes sense to make the API simpler/safer by making it a one time use struct that takes spks in the constructor. @LagginTimes please open a new issue or PR so we can continue exact implementation changes there.

ValuedMammal replied:
On the other hand I think it would make sense to move the logic of `find_base_with` to happen in the constructor, then `get_tip` will only fetch the remote tip and hence could be called multiple times. Just an idea.

Also +1 for single-use workflow.

evanlinjin committed:
feat(bitcoind_rpc): Various fixes for `FilterIter`

* Remove `MAX_REORG_DEPTH` check and error variant. For the scenario
  where the `FilterIter` is of a different network than the receiver, the
  genesis block will be emitted, thus making the receiver error out.
* Fix `Iterator::next` impl so that we only mutate internal state after
  we are certain that no more errors will be returned.
* `Error::NoScripts` is no longer returned since I could not justify a
  rationale for it.
* Refactor `Iterator::next` impl be cleaner.

evanlinjin commented on crates/bitcoind_rpc/src/bip158.rs:
```diff
@@ -33,20 +33,29 @@ pub struct FilterIter<'c, C> {
 
 impl<'c, C: RpcApi> FilterIter<'c, C> {
+    /// Hard cap on how far to walk back when a reorg is detected.
+    const MAX_REORG_DEPTH: u32 = 100;
```
What is the rationale for this check? Are there any downsides of not having this?

100-block reorgs are definitely possible (I.e. black swan events such as WW3). If it's the wrong network, we will emit the genesis block making it impossible to apply to the wallet anyway.

evanlinjin commented on crates/bitcoind_rpc/src/bip158.rs:
```diff
@@ -171,27 +148,80 @@ impl<C: RpcApi> Iterator for FilterIter<'_, C> {
+            // record the scanned block
+            self.blocks.insert(height, hash);
+            // increment best height
+            self.height = height.saturating_add(1);
```
Ideally, we want to deal with errors (esp. IO errors) before we mutate internal state. Otherwise, we may miss a block purely because we dropped connection for one second.

evanlinjin commented on crates/bitcoind_rpc/src/bip158.rs:
```diff
@@ -259,6 +301,7 @@ impl fmt::Display for Error {
             Self::Bip158(e) => e.fmt(f),
             Self::NoScripts => write!(f, "no script pubkeys were provided to match with"),
             Self::Rpc(e) => e.fmt(f),
+            Self::ReorgDepthExceeded => write!(f, "maximum reorg depth exceeded"),
```
Note that this change is breaking (commit message should have `!`). However, I'm proposing to remove this.

evanlinjin reviewed (changes requested):
Thanks for taking this forward.

I addressed my own review comments in commit 30e25d17f8c847039b9412338594f580947f1cba of [`evanlinjin:bdk @ filter_iter`](https://github.com/evanlinjin/bdk/tree/filter_iter). Let me know what you think. Check commit message for rationale.

ValuedMammal reviewed (commented):
I'm not convinced that this represents an objective improvement over Musab's PR.

evanlinjin commented:
> I'm not convinced that this represents an objective improvement over Musab's PR.

@ValuedMammal are you able to provide more context? I'm assuming you are referring to changes in 071fe4034cf45cbaf92eb55921634456099a7f1c?

Here is my justification why these changes are superior:

* **Separation of mutating/non-mutating code -** No state mutations made if anything errors - without this, we may miss blocks if network drops. Better readability.
* **No duplicated logic -** We only need to check whether `prev_blockhash` matches in one place (rather than two places).
* **Remove `MAX_REORG_DEPTH` -** As `MAX_REORG_DEPTH` does not provide any additional security and 100 block reorgs are possible. We are not checking PoW anyway; a malicious node can freely waste our time/return gibberish.

LagginTimes commented:
~~Replaced by #2000.~~

Edit:
Re-opened for further discussion alongside #2000.

evanlinjin commented:
@LagginTimes @ValuedMammal There is no mention of a rationale on why this is closed and replaced by #2000 which is confusing. Let's not do that in the future.

ValuedMammal commented:
When I ran the FilterIter example my canonical txs showed up, but strangely they appeared unconfirmed. I guess because nothing is being retained in `.matched`.

Further discussions surfaced more fundamental issues with the bip158 module which I started to address in #2000, namely the reliability of the internal checkpoint logic.

> `examples/filter_iter.rs`

```rs
    for canon_tx in graph.graph().list_canonical_txs(
        &chain,
        chain.tip().block_id(),
        bdk_chain::CanonicalizationParams::default(),
    ) {
        if !canon_tx.chain_position.is_confirmed() {
            eprintln!(
                "ERROR: canonical txid should be confirmed {}",
                canon_tx.tx_node.txid
            );
        }
    }
```

evanlinjin commented:
@ValuedMammal would you like to be responsible for finishing this work in #2000 and we can close this PR?

LagginTimes commented:
Replaced by #2000.
